### ISSUE TITLE
list_container_objects doesn't work for Azure blob storage under python 2.7.5

### ISSUE THREAD

**Issue by charlesyc** [apache/libcloud#1322]
## Summary

`list_container_objects` doesn't work for Azure blob storage under python 2.7.5

## Detailed Information

###libcloud version:
2.5.0/2.4.0

###Python version:
2.7.5

###Operating System:
RHEL 7.3

###Steps which are needed to reproduce it:
Using the following code to list objects in Azure blob container:
```
from libcloud.storage.types import Provider
from libcloud.storage.providers import get_driver

cls = get_driver(Provider.AZURE_BLOBS)
driver = cls(key='XXX', secret='XXXXXXXX')
container = driver.get_container('myContainer')
objs = driver.list_container_objects(container, '123')

print objs
```

It will throw the following exception at line `objs = driver.list_container_objects(container, '123')`:
```
>>> objs = driver.list_container_objects(container, '123')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/site-packages/libcloud/storage/drivers/azure_blobs.py", line 437, in list_container_objects
    ex_prefix=ex_prefix))
  File "/usr/lib/python2.7/site-packages/libcloud/storage/drivers/azure_blobs.py", line 401, in iterate_container_objects
    params=params)
  File "/usr/lib/python2.7/site-packages/libcloud/common/base.py", line 636, in request
    response = responseCls(**kwargs)
  File "/usr/lib/python2.7/site-packages/libcloud/common/base.py", line 155, in __init__
    self.object = self.parse_body()
  File "/usr/lib/python2.7/site-packages/libcloud/common/azure.py", line 97, in parse_body
    return super(AzureResponse, self).parse_body()
  File "/usr/lib/python2.7/site-packages/libcloud/common/base.py", line 233, in parse_body
    driver=self.connection.driver)
libcloud.common.types.MalformedResponseError: <MalformedResponseException in <libcloud.storage.drivers.azure_blobs.AzureBlobsStorageDriver object at 0x7f9f08838f90> 'Failed to parse XML'>: u'\u010f\u0165\u017c<?xml version="1.0" encoding="utf-8"?><EnumerationResults ContainerName="https://.....
```

The response XML seems to be valid except it begins with `u'\u010f\u0165\u017c`.
I tried in python 2.7.15 and it works fine. Maybe this issue is related to python 2.7.5 only?
Python 2.7.5 is the default version for RHEL 7.3, so please help me. Thanks!

**Comment by c-w** [comment]
Thanks for reporting this and for all the details. As a final piece of information, could you please also provide the subtype/"kind" of the Azure Storage account? You can find this on the storage account's properties page:

![Screenshot showing Azure Storage account subtype](https://user-images.githubusercontent.com/1086421/61712784-1b01aa80-ad25-11e9-9b32-092d8dc9b8c4.png)

**Comment by charlesyc** [comment]
Hi @c-w ,

Thanks for your attention. The "Kind" of my Azure Storage account is "StorageV2 (general purpose v2)
" which is the same one as in your screenshot.
Here is my screenshot:
![image](https://user-images.githubusercontent.com/7031458/61758725-80729b80-adf8-11e9-9090-e8598056b07b.png)

Regards,
Charles

**Comment by charlesyc** [comment]
I finally figured out the root cause is the existing dependency packages of apache-libcloud. Here is what is installed by default of RHEL 7.x:
```
$ pipdeptree -p apache-libcloud
apache-libcloud==2.5.0
  - requests [required: Any, installed: 2.6.0]
```

The dependency package `requests` in pre-installed on RHEL 7.x. But somehow it doesn't work correctly by apache-libcloud.
I used PYTHONPATH to specify the following dependencies manually and it works again:
```
$ pipdeptree -p apache-libcloud
apache-libcloud==2.5.0
  - requests [required: Any, installed: 2.22.0]
    - certifi [required: >=2017.4.17, installed: 2019.6.16]
    - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
    - idna [required: >=2.5,<2.9, installed: 2.8]
    - urllib3 [required: >=1.21.1,<1.26,!=1.25.1,!=1.25.0, installed: 1.25.3]
```
This should be a valid workaround for me. But I am not sure what you should do with the default `requests` package shipped with RHEL 7.x. (It cannot be re-installed pip since it is a distutils installed package)

**Comment by c-w** [comment]
One way to solve this would be to vendor the requests library into libcloud so that we know that there's always a specific version of requests available.

Since we just discovered that libcloud relies on `requests>2.6.0`, another way to at least surface this problem at install time as opposed to run time would be to specify a minimum version of requests that the library requires.

**Comment by Kami** [comment]
I'm not a huge fan of vendoring approach (it could also potentially have licensing implications), but if we identified that it's an issue with the request library version, I'm fine with specifying a minimum version in ``setup.py``.

**Comment by Kami** [comment]
Also, how did you install the library? Did you use pip or a RPM package?

**Comment by charlesyc** [comment]
> Also, how did you install the library? Did you use pip or a RPM package?

The `requests 2.6.0` is pre-installed in RHEL 7.3 OS (probably by RPM). Later I used pip to install libcloud (and all its dependencies) in RHEL 7.3 docker environment (which doesn't have `requests` installed by default). This is how I figured out it is a dependency issue.

**Comment by c-w** [comment]
This is strange. I just pushed a [change](https://github.com/c-w/libcloud-tests/commit/99530c899fd5ae44c7d0c85566eed9e9cffc3f20) to [libcloud-tests](https://github.com/c-w/libcloud-tests) to run the Azure storage integration tests with a specific version of requests and ran the [CI with requests==2.6.0](https://clewolff.visualstudio.com/libcloud-tests/_build/results?buildId=289) but I can't reproduce the error on the Ubuntu/Mac/Windows agents.

I was also unable to reproduce the error on Centos 7 via this approach:

```
$ docker run -it centos:7.3.1611 bash
> curl "https://bootstrap.pypa.io/get-pip.py" -o "get-pip.py"
> python get-pip.py
> pip install requests==2.6.0
> pip install apache-libcloud==2.5.0
> python -c "from libcloud.storage.types import Provider; from libcloud.storage.providers import get_driver; cls = get_driver(Provider.AZURE_BLOBS); driver = cls(key='XXX', secret='YYY); container = driver.get_container('ZZZ'); objs = driver.list_container_objects(container, 'PPP'); print(objs)"
```

So perhaps this is something specific to RHEL's version of requests? Do you have a script I could use to reproduce the error @charlesyc?

**Comment by c-w** [comment]
I did see some failures on [CI using requests<=2.4.0](https://clewolff.visualstudio.com/libcloud-tests/_build/results?buildId=299).

**Comment by charlesyc** [comment]
Hi @c-w :

I think you can reproduce the error by following steps:
1. Pull and run RHEL UBI image (which is RHEL 7.6):
```
docker pull registry.access.redhat.com/ubi7/ubi
docker run -it registry.access.redhat.com/ubi7/ubi
```
2. Install `python-requests` rpm package inside the container:
```
yum install python-requests -y
```
3. Install pip and libcloud:
```
curl -O https://bootstrap.pypa.io/get-pip.py
python get-pip.py
pip install apache-libcloud
```
4. Run the python code in my original post.

Since I am at home, I cannot access to my Azure credentials. Please let me know your result, thanks!

**Comment by c-w** [comment]
Thanks for providing the repro steps. I'm able to reproduce this now. There must be something wrong with the distribution of requests in the yum package:

This doesn't work:

```
yum install python-requests -y
curl -O https://bootstrap.pypa.io/get-pip.py
python get-pip.py
pip install apache-libcloud
python -c "..."  # breaks
```

But this does work:

```
curl -O https://bootstrap.pypa.io/get-pip.py
python get-pip.py
pip install requests==2.6.0
pip install apache-libcloud
python -c "..."  # works
```

**Comment by c-w** [comment]
I diffed the version of requests that's in yum versus pip like so:

```
$ docker run --rm -n rhel -it registry.access.redhat.com/ubi7/ubi
> yum install python-requests -y
$ docker cp rhel:/usr/lib/python2.7/site-packages/requests .
$ docker kill rhel
$ find . -name '*.pyc' -delete; find . -name '*.pyo' -delete
$ git add .; git commit -m "yum install python-requests"
$ docker run --rm --name rhel -it registry.access.redhat.com/ubi7/ubi
> curl -O https://bootstrap.pypa.io/get-pip.py
> python get-pip.py
> pip install requests==2.6.0
$ docker cp rhel:/usr/lib/python2.7/site-packages/requests .
$ find . -name '*.pyc' -delete; find . -name '*.pyo' -delete
$ git add .; git commit -m "pip install requests==2.6.0"
```

Turns out the diff is quite large ([requests-yum-vs-pip.patch](https://github.com/apache/libcloud/files/3437391/requests-yum-vs-pip.patch.txt)):

```
 cacert.pem
 certs.py
 compat.py
 packages/chardet/__init__.py
     ...
 packages/urllib3/__init__.py
     ...
 67 files changed, 17648 insertions(+), 10 deletions(-)
```

I spot checked the diff and the changes in `certs.py` and `compat.py` look fairly innocuous so the difference must be in the version of urllib3 and chardet that are installed by `yum install python-requests -y` vs `pip install requests==2.6.0`.

**Comment by c-w** [comment]
Running this on each of the docker containers:

```py
import requests
print(requests.packages.chardet.__version__)
print(requests.packages.urllib3.__version__)
```

Shows:

| dependency | yum install python-requests | pip install requests==2.6.0 |
| --- | --- | --- |
| chardet |  2.2.1 |  2.3.0 |
| urllib3 | 1.10.2 | 1.10.2 |

Running `pip install chardet==2.3.0` explicitly seems to fix the issue.

**Comment by c-w** [comment]
The real question now is what do we do with this newfound insight... The root cause of this fix is the `python-pip` yum package which should be re-published with `chardet==2.3.0` to match the package on PyPI. Do we want to put an additional assertion in place in libcloud to guard against this situation?
